pyDataverse/api.py

```diff
@@ -2523,17 +2602,41 @@ class SwordApi(Api):
     """
 
     def __init__(
-        self, base_url, api_version="v1.1", api_token=None, sword_api_version="v1.1"
+        self,
+        base_url,
+        api_version="v1.1",
+        api_token=None,
+        sword_api_version="v1.1",
+        *,
+        auth=None,
     ):
         """Init a :class:`SwordApi <pyDataverse.api.SwordApi>` instance.
 
         Parameters
         ----------
         sword_api_version : str
             Api version of Dataverse SWORD API.
+        api_token : str | None
+            An Api token as retrieved from your Dataverse instance.
+        auth : httpx.Auth
+            Note that the SWORD API uses a different authentication mechanism
+            than the native API, in particular it uses `HTTP Basic
+            Authentication
```

Comment:
Thanks for handling the SWORD case!

Reply:
Now that you mention it, I am not even sure this is tested properly.
Maybe I should come up with a test, but I don't know that API, to be honest.

Reply:
Well, it's documented at https://guides.dataverse.org/en/6.3/api/sword.html but I'm happy to answer any questions!

Reply:
Good thing we talked about this.
There was a small test, but I added a few more now to actually test if authentication works:

```python
        def test_sword_api_can_authenticate(self):
            BASE_URL = os.getenv("BASE_URL")
            API_TOKEN = os.getenv("API_TOKEN")
            api = SwordApi(BASE_URL, api_token=API_TOKEN)
            response = api.get_service_document()
            assert response.status_code == 200

        def test_sword_api_cannot_authenticate_without_token(self):
            BASE_URL = os.getenv("BASE_URL")
            api = SwordApi(BASE_URL)
            with pytest.raises(ApiAuthorizationError):
                api.get_service_document()
```

The first test works as expected, but the second does not:

```
./tests/api/test_api.py::TestApiToken::test_sword_api_cannot_authenticate_without_token Failed: [undefined]json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

This is due to an assumption made in `Api._sync_request`: if the resp.status_code is 401, it parses the answer as a JSON:

```python
            if resp.status_code == 401:
                error_msg = resp.json()["message"]
                raise ApiAuthorizationError(
                    "ERROR: HTTP 401 - Authorization error {0}. MSG: {1}".format(
                        kwargs["url"], error_msg
                    )
                )
```

This cannot work with the SWORD case, at it is xml-based as far as I can tell – plus, in this particular case, it only returns an empty response, causing the JSON decoding to fail.
So I guess I will special case that here.

Reply:
I added a try/except around the `resp.json()`.

Reply:
Yes, SWORD is definitely XML-based.

Reply:
And SWORD only supports basic auth. It doesn't support a header or a query parameter.